### PR TITLE
Add 0-mean von Mises sampler to random core

### PR DESCRIPTION
**Pull request by egordon**
Partially addresses question #26987 and issue #11737 by adding a 0-mean von Mises sampler. Transcribes [numpy's implementation](https://github.com/numpy/numpy/blob/b6e8ad83299fc6f87107492df0039dadfb5b9b98/numpy/random/src/distributions/distributions.c#L871), which uses Best and Fisher's algorithm. Different large/small thresholds are used for default 32-bit precision (CDF test still passes at 64-bit precision).

Example code:
```
import jax
import scipy
import matplotlib.pyplot as plt
import numpy as np

angles = np.linspace(-np.pi, np.pi, 100)
kappa = 1.0
loc = 0.
pdf = scipy.stats.vonmises.pdf(angles, loc=loc, kappa=kappa)
samples = jax.random.vonmises(jax.random.key(0), kappa, shape=(10000,))

plt.hist(samples, bins=100, density=True, alpha=0.5, label="PDF")
plt.plot(angles, pdf, label="Samples")
plt.title("Von Mises kappa=1")
plt.legend()
plt.xlabel("Angle")
plt.show()
```

image

(Note: a von Mises Fisher sampler will likely follow soon.)

### CODE REVIEW

File: tests/random_lax_test.py [hunk 1]
```diff
@@ -1389,6 +1392,20 @@ def testRayleigh(self, scale, dtype):
     for samples in [uncompiled_samples, compiled_samples]:
       self._CheckKolmogorovSmirnovCDF(samples, scipy.stats.rayleigh(scale=scale).cdf)
 
+  @jtu.sample_product(
+      kappa = [1e-6, 1e-4, 1e-2, 1e0, 1e2, 1e4, 1e6],
+      dtype = jtu.dtypes.floating)
+  def testVonMises(self, kappa, dtype):
+    key = lambda: self.make_key(0)
+    rand = lambda key: random.vonmises(key, kappa, shape=(10,), dtype=dtype)
```
[the remark this answers is not in the document]
Comment (egordon) [reply tests/random_lax_test.py]: Oops... this was for debugging purposes and not reset.

File: jax/_src/random/core.py [hunk 2]
```diff
@@ -2869,6 +2869,143 @@ def _rayleigh(key, scale, shape, dtype) -> Array:
...
+            -np.pi,
+            np.pi,
+        )
+
+    def mid_kappa_sample(key: Array, kappa: ArrayLike):
+
+        def s_val_from_kappa(kappa: ArrayLike):
+            r_val = 1 + jnp.sqrt(1 + 4 * kappa * kappa)
+            rho_val = (r_val - jnp.sqrt(2 * r_val)) / (2 * kappa)
+            return (1 + rho_val * rho_val) / (2 * rho_val)
```
[the remark this answers is not in the document]
Comment (egordon) [reply jax/_src/random/core.py]: Now called with safe_kappa

File: jax/_src/random/core.py [hunk 3]
```diff
@@ -2869,6 +2869,147 @@ def _rayleigh(key, scale, shape, dtype) -> Array:
   ray = lax.mul(scale, sqrt_u)
   return ray
 
+def vonmises(key: ArrayLike,
+              kappa: RealArray = np.float32(1),
```
Comment (jakevdp) [inline jax/_src/random/core.py]: [nit] match indentation

Also, rather than `np.float32(1)`, please use a scalar `1.0`

File: jax/_src/random/core.py [hunk 4]
```diff
@@ -2869,6 +2869,147 @@ def _rayleigh(key, scale, shape, dtype) -> Array:
...
+  return jnp.reshape(samples, shape)
+
+
+def _vonmises_one(key: Array, kappa: ArrayLike, dtype: DTypeLikeFloat):
+    """Sample from von Mises distribution with mean angle 0 and concentration kappa
+
+    Jax transcription of Numpy's implementation of von Mises sampling, which uses Best and Fisher's algorithm
+    """
+
+    with np.errstate(over="ignore"):
```
Comment (jakevdp) [inline jax/_src/random/core.py]: Since this is executing in a JIT context, I don't think `np.errstate` applies.

File: jax/_src/random/core.py [hunk 5]
```diff
@@ -2869,6 +2869,147 @@ def _rayleigh(key, scale, shape, dtype) -> Array:
...
+  keys = vmap(_split, in_axes=(0, None))(keys, split_count)
+  keys = keys.flatten()
+  kappas = kappa.flatten()
+
+  samples = vmap(partial(_vonmises_one, dtype=dtype))(keys, kappas)
+
+  return jnp.reshape(samples, shape)
+
+
+def _vonmises_one(key: Array, kappa: ArrayLike, dtype: DTypeLikeFloat):
```
Comment (jakevdp) [inline jax/_src/random/core.py]: Use `kappa: Array` here. Given the way this is called, `kappa` will always be an array.

File: jax/_src/random/core.py [hunk 6]
```diff
@@ -2869,6 +2869,147 @@ def _rayleigh(key, scale, shape, dtype) -> Array:
...
+
+    Jax transcription of Numpy's implementation of von Mises sampling, which uses Best and Fisher's algorithm
+    """
+
+    with np.errstate(over="ignore"):
+      kappa_large = jnp.array(1e5).astype(dtype)
+      kappa_small = jnp.array(1e-5).astype(dtype)
+      kappa_mid_small = jnp.array(1e-3).astype(dtype)
+
+    def small_kappa_uniform(key: Array, kappa: ArrayLike):
```
Comment (jakevdp) [inline jax/_src/random/core.py]: `kappa: Array`

File: jax/_src/random/core.py [hunk 7]
```diff
@@ -2869,6 +2869,147 @@ def _rayleigh(key, scale, shape, dtype) -> Array:
...
+  keys = keys.flatten()
+  kappas = kappa.flatten()
+
+  samples = vmap(partial(_vonmises_one, dtype=dtype))(keys, kappas)
+
+  return jnp.reshape(samples, shape)
+
+
+def _vonmises_one(key: Array, kappa: ArrayLike, dtype: DTypeLikeFloat):
+    """Sample from von Mises distribution with mean angle 0 and concentration kappa
```
Comment (jakevdp) [inline jax/_src/random/core.py]: Please use two-space indentation throughout.

File: jax/_src/random/core.py [hunk 8]
```diff
@@ -2869,6 +2869,147 @@ def _rayleigh(key, scale, shape, dtype) -> Array:
...
+            v_val = uniform(vkey, shape=np.shape(kappa), dtype=dtype)
+            return (new_key, kappa, s_val, y_val, v_val, w_val)
+
+        def yw_cond(state):
+            _, _, _, y_val, v_val, _ = state
+            cond1 = y_val * (2.0 - y_val) - v_val >= 0
+            cond2 = jnp.log(y_val / v_val) + 1 - y_val >= 0
+            return ~(cond1 | cond2)
+
+        _, _, _, _, _, w_final = lax_control_flow.while_loop(
```
Comment (jakevdp) [inline jax/_src/random/core.py]: `*_, w_final = `

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/random/core.py (hunk 4 above)
Comment (egordon) [reply jax/_src/random/core.py]: I was getting `RuntimeWarning: overflow encountered in cast` with the `float16` test. I think it applies during compilation.

File: jax/_src/random/core.py [hunk 9]
```diff
@@ -2869,6 +2869,151 @@ def _rayleigh(key, scale, shape, dtype) -> Array:
...
+
+@jit(static_argnums=(2, 3))
+def _vonmises(key, kappa, shape, dtype) -> Array:
+  """Vonmises sampling implementation, with shape and dtype already checked and broadcasted."""
+  kappa = lax.convert_element_type(kappa, dtype)
+  kappa = jnp.broadcast_to(kappa, shape)
+  # split key to match the shape of kappa
+  split_count = math.prod(shape[key.ndim:])
+  keys = key.flatten()
+  keys = vmap(_split, in_axes=(0, None))(keys, split_count)
```
Comment (jakevdp) [inline jax/_src/random/core.py]: Something is strange here: `key` should always have shape `()` (validated by `_check_prng_key` in the calling function). With that in mind, these three lines could be made much simpler.

File: jax/_src/random/core.py [hunk 10]
```diff
@@ -2869,6 +2869,151 @@ def _rayleigh(key, scale, shape, dtype) -> Array:
...
+  keys = keys.flatten()
+  kappas = kappa.flatten()
+
+  samples = vmap(partial(_vonmises_one, dtype=dtype))(keys, kappas)
```
Comment (jakevdp) [inline jax/_src/random/core.py]: This is not the most efficient appraoch for generating random numbers. Rather than vmapping over split keys, you should instead generate an array uniform values from the single input key. Most samplers in this file work similarly (except for `gamma` which is different because it is an iterative sampler)

File: jax/_src/random/core.py [hunk 11]
```diff
@@ -2869,6 +2869,151 @@ def _rayleigh(key, scale, shape, dtype) -> Array:
...
+
+    safe_kappa = jnp.where(kappa < kappa_small, 1.0, kappa)
+
+    def s_val_from_kappa(kappa: Array):
+      safe_kappa = jnp.where(kappa < kappa_mid_small, 1.0, kappa)
+      r_val = 1 + jnp.sqrt(1 + 4 * safe_kappa * safe_kappa)
+      rho_val = (r_val - jnp.sqrt(2 * r_val)) / (2 * safe_kappa)
+      return (1 + rho_val * rho_val) / (2 * rho_val)
+    # Use second order Taylor expansion for small kappa
+    s_val = lax_control_flow.cond(
```
Comment (jakevdp) [inline jax/_src/random/core.py]: In general throughout this implementation you should use `select` rather than `cond`: the batching rule of `cond` is `select`, so the result will be the same in the end.

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/random/core.py [hunk 12]
```diff
@@ -2869,6 +2869,141 @@ def _rayleigh(key, scale, shape, dtype) -> Array:
...
+      rho_val = (r_val - jnp.sqrt(2 * r_val)) / (2 * safe_kappa)
+      return (1 + rho_val * rho_val) / (2 * rho_val)
+    # Use second order Taylor expansion for small kappa
+    s_val = lax.select(
+      safe_kappa < kappa_mid_small, 1.0 / kappa + kappa, s_val_from_kappa(safe_kappa)
+    )
+
+    def body_fn(state):
+      key, accepted, w_out = state
+      new_key, zkey, vkey = _split(key, 3)
```
Comment (jakevdp) [inline jax/_src/random/core.py]: This iterative split approach is problematic, because it's not guaranteed that repeated splits will explore the full key space, and it's possible to get stuck in small cycles that could affect the independence of samples. Instead, you should fold-in the iteration number to generate a subkey.

File: jax/_src/random/core.py [hunk 13]
```diff
@@ -2869,6 +2869,141 @@ def _rayleigh(key, scale, shape, dtype) -> Array:
...
+      accept = cond1 | cond2
+      w_out = lax.select(accept, w_val, w_out)
+      accepted |= accept
+      return (new_key, accepted, w_out)
+
+    def cond_fn(state):
+      accepted = state[-2]
+      return (~accepted).any()
+
+    *_, w_final = lax_control_flow.while_loop(
```
Comment (jakevdp) [inline jax/_src/random/core.py]: `while_loop` can be problematic performance-wise, because the runtime scales as the worst case among the batch of samples and is unbounded. We have some other random samplers that use `while_loop`, and we're looking to move away from those implementations if possible because of these issues – I'm hesitant to add another loop-based sampler. Is there some alternative implementation we could use?

Comment (egordon) [reply jax/_src/random/core.py]: This is what [numpy](https://github.com/numpy/numpy/blob/2f12c608feea5be8f4909f71e8222615c5f8f827/numpy/random/src/distributions/distributions.c#L871) (and by extension scipy) and [cpython](https://github.com/python/cpython/blob/7a468a101268d2b13105f94ae027df8b502d0c87/Lib/random.py#L625) use. 

I could investigate implementing the random mixture method in [Abreu 2008](https://doi.org/10.1109%2FTCOMM.2008.060510), but that would be novel and might be a bit beyond the time I can reasonably dedicate.

File: jax/_src/random/core.py (hunk 12 above)
Comment (egordon) [reply jax/_src/random/core.py]: Would a fold-in then split be sufficient? 

I.e.
`new_key, zkey, vkey = _split(fold_in(key, iter_idx), 3)`

If so, do you want me to also update the other distributions that do iterative split (e.g. poisson) or leave it for a different PR?

File: jax/_src/random/core.py (hunk 13 above)
Comment (jakevdp) [reply jax/_src/random/core.py]: NumPy and SciPy have different constraints though: e.g. on a CPU, it's trivial to run a `while_loop` for a different number of iterations on each value of the input array. On accelerators that can't do fast scalar ops (like GPU and TPU), that approach doesn't work, and so in practice this means we must do the worst-cast `while_loop` over the entire batch. For algorithms with a large skew in the required number of iterations (like `gamma`) this leads to very slow code.

I'd like to avoid adding more very slow code to `jax.random`, because it's not a great experience for users.

Comment (egordon) [reply jax/_src/random/core.py]: Understood.

It does look like even [torch](https://github.com/pytorch/pytorch/blob/7d8146346b482852b8e30cfc5f1e8f938c2ce572/torch/distributions/von_mises.py#L93) uses this approach, so it seems standard, but that doesn't mean jax can't be better :)

I'll address your other comment but I can take a couple cycles to see if I have time to implement Abreu if it is better.

As an aside, if I do PR von mises Fisher, I have not seen an algorithm for dim>3 that doesn't use rejection sampling. So it may be unavoidable if you want that distribution as well.

Comment (egordon) [reply jax/_src/random/core.py]: Unless I try a table lookup or polynomial fit approach for some of the intermediate values, Abreu requires iv (up to v=10).

#2466

Comment (egordon) [reply jax/_src/random/core.py]: Actually since I only need integer order I can just compute iv iteratively from i0 and i1.

Let me try that.

Comment (egordon) [reply jax/_src/random/core.py]: I think this will require a bit too much time to implement.

The (or at least *a*) problem is that the computation of `mu_T=I_v(kappa)/I_0(kappa)` is just not numerically stable when done naively with the recursive relationship. For kappa as small as 1e2, i0 is inf for `float32`. And even for `kappa=1.0`, only the first 3 orders are actually close to the `scipy` implementation for `float32`, and only 9 orders are close for `float64`.

If you don't feel comfortable merging Best and Fisher and maybe making Abreu an issue conditioned on a stable implementation of integer I_v/I_0, then I don't think this PR will land.
